File: ctrlfile/src/parser.zig
const std = @import("std");
const lex = @import("lexer.zig");
const ast = @import("ast.zig");

pub const ParseError = error{
    UnexpectedToken,
    OutOfMemory,
};

pub const Parser = struct {
    allocator: std.mem.Allocator,
    tokens: []lex.Token,
    pos: usize = 0,

    pub fn init(allocator: std.mem.Allocator, tokens: []lex.Token) Parser {
        return Parser{ .allocator = allocator, .tokens = tokens };
    }

    fn peek(self: *Parser) lex.Token {
        return self.tokens[self.pos];
    }

    fn advance(self: *Parser) lex.Token {
        const t = self.tokens[self.pos];
        if (self.pos + 1 < self.tokens.len) self.pos += 1;
        return t;
    }

    fn expect(self: *Parser, kind: lex.TokenType) ParseError!lex.Token {
        const t = self.peek();
        if (t.kind != kind) {
            std.debug.print("linia {d}: nieoczekiwany token '{s}'\n", .{ t.line, t.text });
            return ParseError.UnexpectedToken;
        }
        return self.advance();
    }

    fn check(self: *Parser, kind: lex.TokenType) bool {
        return self.peek().kind == kind;
    }

    pub fn parseProgram(self: *Parser) ParseError!ast.Program {
        var runs = std.ArrayList(ast.RunBlock).init(self.allocator);
        var whens = std.ArrayList(ast.WhenBlock).init(self.allocator);

        while (!self.check(.eof)) {
            if (self.check(.kw_run)) {
                try runs.append(try self.parseRunBlock());
            } else if (self.check(.kw_when)) {
                try whens.append(try self.parseWhenBlock());
            } else {
                std.debug.print("linia {d}: oczekiwano 'run' lub 'when', jest '{s}'\n", .{ self.peek().line, self.peek().text });
                return ParseError.UnexpectedToken;
            }
        }

        return ast.Program{
            .runs = try runs.toOwnedSlice(),
            .whens = try whens.toOwnedSlice(),
        };
    }

    fn parseRunBlock(self: *Parser) ParseError!ast.RunBlock {
        _ = try self.expect(.kw_run);
        const name_tok = try self.expect(.string);
        _ = try self.expect(.lbrace);

        var steps = std.ArrayList(ast.Step).init(self.allocator);
        while (!self.check(.rbrace)) {
            try steps.append(try self.parseStep());
        }
        _ = try self.expect(.rbrace);

        return ast.RunBlock{
            .name = name_tok.text,
            .steps = try steps.toOwnedSlice(),
        };
    }

    fn parseStep(self: *Parser) ParseError!ast.Step {
        _ = try self.expect(.kw_step);
        const num_tok = try self.expect(.number);
        _ = try self.expect(.colon);

        var actions = std.ArrayList(ast.Action).init(self.allocator);
        try actions.append(try self.parseAction());
        while (self.check(.amp)) {
            _ = self.advance();
            try actions.append(try self.parseAction());
        }

        const number = std.fmt.parseInt(u32, num_tok.text, 10) catch 0;
        return ast.Step{
            .number = number,
            .actions = try actions.toOwnedSlice(),
        };
    }

    fn parseAction(self: *Parser) ParseError!ast.Action {
        const name_tok = try self.expect(.ident);
        _ = try self.expect(.lparen);
        const arg_tok = try self.expect(.string);
        _ = try self.expect(.rparen);

        var kind: ast.ActionKind = .cmd;
        if (std.mem.eql(u8, name_tok.text, "install")) {
            kind = .install;
        } else if (!std.mem.eql(u8, name_tok.text, "cmd")) {
            std.debug.print("linia {d}: nieznana akcja '{s}'\n", .{ name_tok.line, name_tok.text });
            return ParseError.UnexpectedToken;
        }

        return ast.Action{ .kind = kind, .value = arg_tok.text };
    }

    fn parseWhenBlock(self: *Parser) ParseError!ast.WhenBlock {
        _ = try self.expect(.kw_when);
        const condition = try self.parseExpr();
        _ = try self.expect(.lbrace);

        var actions = std.ArrayList(ast.Action).init(self.allocator);
        while (!self.check(.rbrace)) {
            try actions.append(try self.parseAction());
        }
        _ = try self.expect(.rbrace);

        return ast.WhenBlock{
            .condition = condition,
            .actions = try actions.toOwnedSlice(),
        };
    }

    fn newExpr(self: *Parser, e: ast.Expr) ParseError!*ast.Expr {
        const p = try self.allocator.create(ast.Expr);
        p.* = e;
        return p;
    }

    fn parseExpr(self: *Parser) ParseError!*ast.Expr {
        return self.parseTernary();
    }

    fn parseTernary(self: *Parser) ParseError!*ast.Expr {
        const cond = try self.parseLogicOr();
        if (self.check(.question)) {
            _ = self.advance();
            const then_e = try self.parseExpr();
            _ = try self.expect(.colon);
            const else_e = try self.parseExpr();
            return self.newExpr(ast.Expr{
                .kind = .ternary,
                .cond = cond,
                .then_expr = then_e,
                .else_expr = else_e,
            });
        }
        return cond;
    }

    fn parseLogicOr(self: *Parser) ParseError!*ast.Expr {
        var left = try self.parseLogicAnd();
        while (self.check(.oror)) {
            _ = self.advance();
            const right = try self.parseLogicAnd();
            left = try self.newExpr(ast.Expr{ .kind = .or_, .left = left, .right = right });
        }
        return left;
    }

    fn parseLogicAnd(self: *Parser) ParseError!*ast.Expr {
        var left = try self.parseEquality();
        while (self.check(.andand)) {
            _ = self.advance();
            const right = try self.parseEquality();
            left = try self.newExpr(ast.Expr{ .kind = .and_, .left = left, .right = right });
        }
        return left;
    }

    fn parseEquality(self: *Parser) ParseError!*ast.Expr {
        var left = try self.parseUnary();
        while (self.check(.eqeq) or self.check(.noteq)) {
            const op = self.advance();
            const right = try self.parseUnary();
            const kind: ast.ExprKind = if (op.kind == .eqeq) .eq else .neq;
            left = try self.newExpr(ast.Expr{ .kind = kind, .left = left, .right = right });
        }
        return left;
    }

    fn parseUnary(self: *Parser) ParseError!*ast.Expr {
        if (self.check(.not)) {
            _ = self.advance();
            const inner = try self.parseUnary();
            return self.newExpr(ast.Expr{ .kind = .not_, .left = inner });
        }
        return self.parsePrimary();
    }

    fn parsePrimary(self: *Parser) ParseError!*ast.Expr {
        if (self.check(.lparen)) {
            _ = self.advance();
            const inner = try self.parseExpr();
            _ = try self.expect(.rparen);
            return inner;
        }
        if (self.check(.string)) {
            const t = self.advance();
            return self.newExpr(ast.Expr{ .kind = .string, .str = t.text });
        }
        if (self.check(.ident)) {
            const name_tok = self.advance();
            _ = try self.expect(.lparen);
            const arg_tok = try self.expect(.string);
            _ = try self.expect(.rparen);
            return self.newExpr(ast.Expr{
                .kind = .call,
                .call_name = name_tok.text,
                .call_arg = arg_tok.text,
            });
        }
        std.debug.print("linia {d}: nieoczekiwany token w wyrazeniu '{s}'\n", .{ self.peek().line, self.peek().text });
        return ParseError.UnexpectedToken;
    }
};

File: ctrlfilesafe/src/parser.rs
use crate::ast::{Action, Expr, Program, RunBlock, Step, WhenBlock};
use crate::lexer::{Token, TokenKind};

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn advance(&mut self) -> Token {
        let t = self.tokens[self.pos].clone();
        if self.pos + 1 < self.tokens.len() {
            self.pos += 1;
        }
        t
    }

    fn check(&self, kind: TokenKind) -> bool {
        self.peek().kind == kind
    }

    fn check_ident(&self, text: &str) -> bool {
        self.peek().kind == TokenKind::Ident && self.peek().text == text
    }

    fn expect(&mut self, kind: TokenKind) -> Result<Token, String> {
        if self.peek().kind != kind {
            return Err(format!(
                "linia {}: oczekiwano {:?}, jest '{}'",
                self.peek().line,
                kind,
                self.peek().text
            ));
        }
        Ok(self.advance())
    }

    pub fn parse_program(&mut self) -> Result<Program, String> {
        let mut program = Program::default();
        while !self.check(TokenKind::Eof) {
            if self.check(TokenKind::KwRun) {
                program.runs.push(self.parse_run_block()?);
            } else if self.check(TokenKind::KwWhen) {
                program.whens.push(self.parse_when_block()?);
            } else {
                return Err(format!(
                    "linia {}: oczekiwano 'run' lub 'when', jest '{}'",
                    self.peek().line,
                    self.peek().text
                ));
            }
        }
        Ok(program)
    }

    fn parse_run_block(&mut self) -> Result<RunBlock, String> {
        self.expect(TokenKind::KwRun)?;
        let name = self.expect(TokenKind::Str)?.text;

        let mut needs = Vec::new();
        if self.check_ident("needs") {
            self.advance();
            needs.push(self.expect(TokenKind::Str)?.text);
            while self.check(TokenKind::Comma) {
                self.advance();
                needs.push(self.expect(TokenKind::Str)?.text);
            }
        }

        self.expect(TokenKind::LBrace)?;
        let mut steps = Vec::new();
        while !self.check(TokenKind::RBrace) {
            steps.push(self.parse_step()?);
        }
        self.expect(TokenKind::RBrace)?;

        Ok(RunBlock { name, needs, steps })
    }

    fn parse_step(&mut self) -> Result<Step, String> {
        self.expect(TokenKind::KwStep)?;
        let num_tok = self.expect(TokenKind::Number)?;
        self.expect(TokenKind::Colon)?;

        let mut actions = vec![self.parse_action()?];
        while self.check(TokenKind::Amp) {
            self.advance();
            actions.push(self.parse_action()?);
        }

        let number: u32 = num_tok.text.parse().unwrap_or(0);
        Ok(Step { number, actions })
    }

    fn parse_action(&mut self) -> Result<Action, String> {
        let name_tok = self.expect(TokenKind::Ident)?;
        self.expect(TokenKind::LParen)?;

        let action = match name_tok.text.as_str() {
            "cmd" => {
                self.expect(TokenKind::LBracket)?;
                let mut argv = vec![self.expect(TokenKind::Str)?.text];
                while self.check(TokenKind::Comma) {
                    self.advance();
                    argv.push(self.expect(TokenKind::Str)?.text);
                }
                self.expect(TokenKind::RBracket)?;
                Action::Cmd(argv)
            }
            "shell" => {
                let s = self.expect(TokenKind::Str)?.text;
                Action::Shell(s)
            }
            "install" => {
                let s = self.expect(TokenKind::Str)?.text;
                Action::Install(s)
            }
            other => {
                return Err(format!(
                    "linia {}: nieznana akcja '{}'",
                    name_tok.line, other
                ))
            }
        };

        self.expect(TokenKind::RParen)?;
        Ok(action)
    }

    fn parse_when_block(&mut self) -> Result<WhenBlock, String> {
        self.expect(TokenKind::KwWhen)?;
        let condition = self.parse_expr()?;
        self.expect(TokenKind::LBrace)?;
        let mut actions = Vec::new();
        while !self.check(TokenKind::RBrace) {
            actions.push(self.parse_action()?);
        }
        self.expect(TokenKind::RBrace)?;
        Ok(WhenBlock { condition, actions })
    }

    fn parse_expr(&mut self) -> Result<Expr, String> {
        self.parse_ternary()
    }

    fn parse_ternary(&mut self) -> Result<Expr, String> {
        let cond = self.parse_or()?;
        if self.check(TokenKind::Question) {
            self.advance();
            let then_e = self.parse_expr()?;
            self.expect(TokenKind::Colon)?;
            let else_e = self.parse_expr()?;
            return Ok(Expr::Ternary(Box::new(cond), Box::new(then_e), Box::new(else_e)));
        }
        Ok(cond)
    }

    fn parse_or(&mut self) -> Result<Expr, String> {
        let mut left = self.parse_and()?;
        while self.check(TokenKind::OrOr) {
            self.advance();
            let right = self.parse_and()?;
            left = Expr::Or(Box::new(left), Box::new(right));
        }
        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Expr, String> {
        let mut left = self.parse_equality()?;
        while self.check(TokenKind::AndAnd) {
            self.advance();
            let right = self.parse_equality()?;
            left = Expr::And(Box::new(left), Box::new(right));
        }
        Ok(left)
    }

    fn parse_equality(&mut self) -> Result<Expr, String> {
        let mut left = self.parse_unary()?;
        loop {
            if self.check(TokenKind::EqEq) {
                self.advance();
                let right = self.parse_unary()?;
                left = Expr::Eq(Box::new(left), Box::new(right));
            } else if self.check(TokenKind::NotEq) {
                self.advance();
                let right = self.parse_unary()?;
                left = Expr::Neq(Box::new(left), Box::new(right));
            } else {
                break;
            }
        }
        Ok(left)
    }

    fn parse_unary(&mut self) -> Result<Expr, String> {
        if self.check(TokenKind::Not) {
            self.advance();
            let inner = self.parse_unary()?;
            return Ok(Expr::Not(Box::new(inner)));
        }
        self.parse_primary()
    }

    fn parse_primary(&mut self) -> Result<Expr, String> {
        if self.check(TokenKind::LParen) {
            self.advance();
            let inner = self.parse_expr()?;
            self.expect(TokenKind::RParen)?;
            return Ok(inner);
        }
        if self.check(TokenKind::Str) {
            let t = self.advance();
            return Ok(Expr::Str(t.text));
        }
        if self.check(TokenKind::Ident) {
            let name_tok = self.advance();
            self.expect(TokenKind::LParen)?;
            let arg_tok = self.expect(TokenKind::Str)?;
            self.expect(TokenKind::RParen)?;
            return Ok(Expr::Call { name: name_tok.text, arg: arg_tok.text });
        }
        Err(format!(
            "linia {}: nieoczekiwany token w wyrazeniu '{}'",
            self.peek().line,
            self.peek().text
        ))
    }
}
